server/ServerApp.cpp

```diff
@@ -2288,36 +2288,42 @@ namespace {
         }
     }
 
-    void GetFleetsVisibleToEmpireAtSystem(std::set<int>& visible_fleets, int empire_id, int system_id,
-                                          const ScriptingContext& context)
+    bool HasBlockadeOrGetFleetsVisibleToEmpireAtSystem(std::set<int>& visible_fleets, int empire_id,
+                                                       int system_id, const ScriptingContext& context)
     {
         const ObjectMap& objects{context.ContextObjects()};
 
         visible_fleets.clear();
         auto* system = objects.getRaw<System>(system_id);
         if (!system)
-            return; // no such system
+            return false; // no such system
         const auto& fleet_ids = system->FleetIDs();
         if (fleet_ids.empty())
-            return; // no fleets to be seen
+            return false; // no fleets to be seen
         if (empire_id != ALL_EMPIRES && !context.GetEmpire(empire_id))
-            return; // no such empire
+            return false; // no such empire
 
-        TraceLogger(combat) << "\t** GetFleetsVisibleToEmpire " << empire_id << " at system " << system->Name();
+        TraceLogger(combat) << "\t** HasBlockadeOrGetFleetsVisibleToEmpireAtSystem " << empire_id << " at system " << system->Name();
+        // first pass over all fleets, to find a blockade and handle empire visibility
         // for visible fleets by an empire, check visibility of fleets by that empire
-        if (empire_id != ALL_EMPIRES) {
-            for (const auto* fleet : objects.findRaw<Fleet>(fleet_ids)) {
-                if (!fleet)
-                    continue;
-                if (fleet->OwnedBy(empire_id))
-                    continue;   // don't care about fleets owned by the same empire for determining combat conditions
-                Visibility fleet_vis = context.ContextVis(fleet->ID(), empire_id);
-                TraceLogger(combat) << "\t\tfleet (" << fleet->ID() << ") has visibility rank " << fleet_vis;
-                if (fleet_vis >= Visibility::VIS_BASIC_VISIBILITY)
-                    visible_fleets.emplace(fleet->ID());
+        for (const auto* fleet : objects.findRaw<Fleet>(fleet_ids)) {
+            if (!fleet)
+                continue;
+            if (fleet->Blockaded(context)) {
+                TraceLogger(combat) << "\t\tfleet (" << fleet->ID() << ") is blockaded. Shortcut determining visibility. ";
+                return true; // found a blockade -> combat ensues
```

Comment:
this seems problematic... you're returning true to indicate there is a blockade in a system if you find any fleet that is blockaded. But the blockaded and blockading fleets might not want to initiate combat in that case (unless one or both are aggressive).

Reply:
also, this presumably won't catch cases where an unarmed or non-aggressive ship is blockaded. the blockader won't initiate combat, and the blockaded ship won't be considered aggressive. probably the blockading ship should still be made visible to the blockaded empire in that case?

Reply:
> Seemingly the blockading fleet should still be made visible to the blockaded empire in that case, though?

I am not convinced it should. If we do that disrupting supply should also make the blockader visible for consistency.

> That could lead to a bit of an inconsistency, in that when a blockade of an aggressive armed fleet causes combat  to happen, the blockading ships potentially initially not visible (depending on their stealth), and any ship that  doesn't attack during the combat wouldn't be revealed by it. But if a non-aggressive/non-armed fleet is blockaded by an obstrutive fleet, and no combat happens, and (somehow) the blockading fleet is thus revealed, presumably the whole blockadaing fleet would be made visible...

maybe only one (random) blocking ship would need to be made visible if non-military ships/supply or obstructed

For the moment (i.e. this PR) I would not change this, rather start discussion/making a use case matrix.

That one is not able to break a hidden blockade at all is almost game breaking - i think it is as bad as the invincible hidden carriers we had once.